src/communication/in_memory_rpc_server.rs

```diff
@@ -0,0 +1,807 @@
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use protobuf::Message;
+use tracing::debug;
+
+use crate::{
+    LocalUriProvider, UAttributes, UAttributesError, UAttributesValidators, UCode, UListener,
+    UMessage, UMessageBuilder, UMessageType, UPayloadFormat, UPriority, UStatus, UTransport, UUri,
+};
+
+use super::{RegistrationError, RequestHandler, RpcServer, ServiceInvocationError, UPayload};
+
+struct RequestListener {
+    request_handler: Arc<dyn RequestHandler>,
+    transport: Arc<dyn UTransport>,
+}
+
+impl RequestListener {
+    async fn process_valid_request(&self, request_message: UMessage) {
+        let Some(resource_id) = request_message
+            .attributes
+            .as_ref()
+            .and_then(|attribs| attribs.sink.as_ref())
+            .and_then(|uri| u16::try_from(uri.resource_id).ok())
+        else {
+            // the conversion cannot fail because the UListener has already verified that the
+            // request message is indeed a valid uProtocol RPC Request with a proper sink
+            // URI representing a method (having a 16 bit resource ID).
+            return;
+        };
+
+        let transport_clone = self.transport.clone();
+        let request_handler_clone = self.request_handler.clone();
+
+        let request_id = request_message
+            .attributes
+            .get_or_default()
+            .id
+            .get_or_default();
+        let request_timeout = request_message
+            .attributes
+            .get_or_default()
+            .ttl
+            .unwrap_or(10_000);
+        let payload = request_message.payload;
+        let payload_format = request_message
+            .attributes
+            .get_or_default()
+            .payload_format
+            .enum_value_or_default();
+        let request_payload = payload.map(|data| UPayload::new(data, payload_format));
+
+        debug!(ttl = request_timeout, id = %request_id, "processing RPC request");
+
+        let invocation_result_future =
+            request_handler_clone.invoke_method(resource_id, request_payload);
+        let outcome = tokio::time::timeout(
+            Duration::from_millis(request_timeout as u64),
+            invocation_result_future,
+        )
+        .await
+        .map_err(|_e| {
+            debug!(ttl = request_timeout, "request handler timed out");
```

Comment:
Well, the RpcServer does not really experience an erroneous situation, does it? We are expecting this to happen and we are handling it appropriately. No service technician needs to be alarmed about this happening, right?

Reply:
I agree that this might not need to be treated as an error, but perhaps we can use info or something else here? At least, we can filter other debug messages if we only care about timeout issues (or something similar).
It would be great if we could have different levels of messages in the code instead of only using debug.

Reply:
I originally added the log statements in preparation for supporting real distributed tracing. I guess we should distinguish between logging a statement to e.g. the file system and marking the current span as erroneous. The former will most likely be used during development to see what is going on locally. The latter can be used to figure out, where in the distributed system the processing of a message has actually failed.

That said, I will need to dig a little deeper into the *tracing* crate in order to figure out, what meaning the *level* has when adding info to a span. For the time being, I guess I am ok with raising the log level to `INFO` here :-)